### PR TITLE
formula: add on_linux and on_macos blocks

### PR DESCRIPTION
**Pull request by iMichka**
for install and others

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

### CODE REVIEW

File: Library/Homebrew/extend/os/linux/formula.rb [hunk 1]
```diff
@@ -1,6 +1,12 @@
 # frozen_string_literal: true
 
 class Formula
+  undef on_linux
+
+  def on_linux(&_block)
+    yield
+  end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/formula.rb:8]:
```suggestion
  def on_linux(&block)
    yield if block_given?
  end
```

File: Library/Homebrew/extend/os/mac/formula.rb [hunk 2]
```diff
@@ -1,6 +1,12 @@
 # frozen_string_literal: true
 
 class Formula
+  undef on_macos
+
+  def on_macos(&_block)
+    yield
+  end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/mac/formula.rb:8]:
```suggestion
  def on_macos(&block)
    yield if block_given?
  end
```

File: Library/Homebrew/test/cmd/install_spec.rb [hunk 3]
```diff
@@ -73,4 +73,27 @@ def install
...
+          raise
+        end
+
+        on_linux do
+          prefix.install Dir["*"]
+        end
+      end
+    RUBY
+
+    expect { brew "install", "testball1" }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/cmd/install_spec.rb:93]: This shouldn't be an integration test but a `formula_spec` test instead.

Comment (iMichka) [reply Library/Homebrew/test/cmd/install_spec.rb:93]: I tried that initially  but I did not find any `install` calls in `formula_spec`. I need to dig further if that is possible there.

Comment (MikeMcQuaid) [reply Library/Homebrew/test/cmd/install_spec.rb:93]: Look for references to `f.install` and `f.brew { f.install }`

File: Library/Homebrew/extend/os/mac/formula.rb (hunk 2 above)
Comment (reitermarkus) [reply Library/Homebrew/extend/os/mac/formula.rb:8]: I think these should also raise an error if no block is given.

Comment (iMichka) [reply Library/Homebrew/extend/os/mac/formula.rb:8]: We did not do that for the on_linux/on_macos block below. Should we be consistent? If yes, should I do this in this PR?

Comment (MikeMcQuaid) [reply Library/Homebrew/extend/os/mac/formula.rb:8]: > Should we be consistent? If yes, should I do this in this PR?

Yes and yes, please!

File: Library/Homebrew/formula.rb [hunk 4]
```diff
@@ -1366,6 +1366,18 @@ def inspect
     "#<Formula #{name} (#{active_spec_sym}) #{path}>"
   end
 
+  # Block only executed on macOS. No-op on Linux.
+  # <pre>on_macos do
+  #   depends_on "mac_only_dep"
+  # end</pre>
+  def on_macos(&_block); end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:1373]: These could also probably raise if no block given?

File: Library/Homebrew/extend/os/linux/formula.rb [hunk 5]
```diff
@@ -2,6 +2,14 @@
 # frozen_string_literal: true
 
 class Formula
+  undef on_linux
+
+  def on_linux(&_block)
+    yield if block_given?
+
+    raise "No block content defined for on_linux block" unless block_given?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/formula.rb:10]:
```suggestion
    if block_given?
      yield
    else
      raise "No block content defined for on_linux block"
    end
```

or 

```suggestion
    raise "No block content defined for on_linux block" unless block_given?

    yield
```

**Comment by iMichka** [comment]
@MikeMcQuaid I made the asked changes but the style check is not happy with that:
```
Library/Homebrew/extend/os/linux/formula.rb:8:5: C: Style/GuardClause: Use a guard clause (raise "No block content defined for on_linux block" unless block_given?) instead of wrapping the code inside a conditional expression.
    if block_given?
    ^^
```

I'll change it make it pass :)

**Comment by MikeMcQuaid** [comment]
> This pull request is still a work in progress
> Draft pull requests cannot be merged.

@iMichka Is this still true? If not: can you mark as "ready for review"?

**Comment by iMichka** [comment]
> @iMichka Is this still true? If not: can you mark as "ready for review"?

Done. Forgot to remove it.

File: Library/Homebrew/extend/os/linux/formula.rb (hunk 5 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/formula.rb:10]:
```suggestion
    raise "No block content defined for on_linux block" unless block_given?

    yield
```

File: Library/Homebrew/extend/os/linux/formula.rb [hunk 6]
```diff
@@ -12,7 +20,9 @@ class << self
     undef on_linux
 
     def on_linux(&_block)
-      yield
+      yield if block_given?
+
+      raise "No block content defined for on_linux block" unless block_given?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/formula.rb:25]:
```suggestion
      raise "No block content defined for on_linux block" unless block_given?

      yield
```

File: Library/Homebrew/extend/os/mac/formula.rb [hunk 7]
```diff
@@ -2,11 +2,21 @@
 # frozen_string_literal: true
 
 class Formula
+  undef on_macos
+
+  def on_macos(&_block)
+    yield if block_given?
+
+    raise "No block content defined for on_macos block" unless block_given?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/mac/formula.rb:10]:
```suggestion
    raise "No block content defined for on_macos block" unless block_given?

    yield
```

File: Library/Homebrew/extend/os/mac/formula.rb [hunk 8]
```diff
@@ -2,11 +2,21 @@
...
+  end
+
   class << self
     undef on_macos
 
     def on_macos(&_block)
-      yield
+      yield if block_given?
+
+      raise "No block content defined for on_macos block" unless block_given?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/mac/formula.rb:19]:
```suggestion
      raise "No block content defined for on_macos block" unless block_given?

      yield
```

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**